src/python/pants/backend/jvm/tasks/jvm_compile/jvm_compile.py

```diff
@@ -582,8 +586,38 @@ def execute(self):
                             classpath_product.remove_for_target(cc.target, [(conf, cc.classes_dir)])
                             classpath_product.add_for_target(cc.target, [(conf, cc.jar_file)])
 
-        if dependees_of_target_roots is not None:
-            self.create_extra_products_for_targets(dependees_of_target_roots)
+        if modulizable_targets is not None:
+            compilable_modulizable_targets = set(
+                filter(lambda x: self.select(x), modulizable_targets)
+            )
+            self.create_extra_products_for_targets(compilable_modulizable_targets)
+
+    def calculate_jvm_modulizable_targets(self):
+        def is_jvm_or_resource_target(t):
+            return isinstance(t, (JvmTarget, JvmApp, JarLibrary, Resources))
+
+        jvm_and_resources_target_roots = set(
+            filter(is_jvm_or_resource_target, self.context.target_roots)
+        )
+        jvm_and_resources_target_roots_minus_thrift_addresses = set(
+            t.address for t in filter(lambda x: not x.is_synthetic, jvm_and_resources_target_roots)
+        )
+        all_targets = set(self.context.targets())
+        modulizable_targets = set(
+            t
+            for t in self.context.build_graph.transitive_dependees_of_addresses(
+                jvm_and_resources_target_roots_minus_thrift_addresses,
+                predicate=lambda x: x in all_targets,
+            )
+            if is_jvm_or_resource_target(t)
+        )
+        synthetic_modulizable_targets = set(filter(lambda x: x.is_synthetic, modulizable_targets))
+        if len(synthetic_modulizable_targets) > 0:
```

Comment:
Its not clear to me why removing synthetic target roots before mean that no synthetic target can be a dependee of another target root.

Reply:
This is because we want to freeze all synthetic targets into jars. If there is any `synthetic_modulizable_targets`, then the module graph in IDE will be broken because jars cannot depend back onto a source module.